### ISSUE TITLE
Cannot find namespace 'JSX'

### ISSUE THREAD

**Issue by Ruluk** [planttheidea/moize#186]
Hey,

In TypeScript, the `index.d.ts` file includes the following lines (15-19):

```typescript
export type FunctionalComponent<Props extends object> = ((
    props: Props
) => JSX.Element) & {
    displayName?: string;
};
```

This seems to raise an error when building a TypeScript project that doesn't have React:

```
error TS2503: Cannot find namespace 'JSX'.
```

Effectively, adding a global declaration fixed the issue for me:
```typescript
declare global {
  namespace JSX {
    type Element = any;
  }
}
```

But you may want to add a proper fix in the declaration file.

**Comment by planttheidea** [comment]
The `JSX` namespace is something automatically provided by `@types/react`. Likely this is occurring because `@types/react` is not declared as an explicit dependency, and does not exist in your project.

If I spin up a beta with the fix, would you be willing to install it and confirm it solves the issue?

**Comment by planttheidea** [comment]
It should be available now as `6.1.5-beta.0`, which can be installed via `npm install moize@next`.

If I don't hear back in the next day I'll go ahead and publish because I'm pretty sure this will resolve the issue, just wanted to be sure.

**Comment by Ruluk** [comment]
Hey. Thanks for the prompt reply!

This did not work, though. The internal dependencies won't work. They might work if you request a peer dependency, but the best solution in this case may be explicitly importing `Element` from `@types/react` in your declaration file.

**Comment by planttheidea** [comment]
It is possible that the original approach didn't work because you did not explicitly include `react` types in your `tsconfig`, and therefore doesn't exist in the project. Frankly, consumers shouldn't need to, so in hindsight this was a naive approach.

Alright, take two ... I've released a new prepatch (`6.1.5-beta.1`) which eliminates the reliance on the external React type entirely in favor of an internal type. Considering the optionality of React usage with the library, this seems to be a more appropriate approach, and the maintenance cost no larger because of the internals I already have related to React elements.

This _should_ solve the issue, but again would appreciate you giving it a go. If this doesn't do it, I'll need to rethink things.

**Comment by benvd** [comment]
Not OP, but `6.1.5-beta.1` fixes the issue for me.

**Comment by Ruluk** [comment]
It works for me too. Thanks one more time for the prompt replies and for the fix!

**Comment by Ruluk** [comment]
One thing to point out, though. Isn't there a chance that this could break the typing for somebody else that _does_ use React?

**Comment by planttheidea** [comment]
@Ruluk unlikely - the type has the same shape as the `JSX.Element` type, but with additional widening because I didn't want to reproduce the narrower internals that React has.

That said ... Since we can guarantee the `type` of the element is a function, I can probably just use `ReturnType<Passed component>` to maintain the narrowing. I'll give this a try.

**Comment by planttheidea** [comment]
Alright, I made the above change leveraging `ReturnType<>`, it looks good. Turns out it didn't actually make a difference for consumers one way or the other (just internal typing), but for posterity and transparency here are the details.

Other than the internal typing for a moized component, it turns out the exposed use of `JSX.Element` was for an exposed type `FunctionalComponent`. This was only used for an arcane stats typing, and it turns out that inference could be used for what it was originally intended for. Because of this, I retained the re-implementation of `JSX.Element` internally for typing of the return type of `FunctionalComponent`, but deprecated `FunctionalComponent` and removed all uses other than the exposed typing. I only left it in to avoid breaking changes, in case someone somewhere was using it in their code. Even if they were, it should not break anything, and there is now guidance about leaning into inference instead of manually using `FunctionalComponent` typing.

I've released `6.1.5` with these changes, but let me know if there are any more issues!
